### PR TITLE
Added Optional emptiness handler

### LINKED ISSUE #274: Proposal: Understand Optional isPresent/isEmpty/isAbsent
Right now NullAway can understand cases like AutoValue where 

```java
if (a.foo() != null) {
  a.foo().bar();
}
```

I propose adding support for an `Optional` analogue:

```java
if (fooOptional.isPresent()) {
  fooOptional.get().bar(); // Implicitly understood to be safe
}
```

I think this is a good fit for NullAway since `Optional` is Java 8/Guava's alternative to `null` in common cases, and has precedent for this kind of data flow awareness of "was null-checked already". This would just be a different flavor.

### PR DESCRIPTION
**Pull request by shas19**
Added Optional Emptiness handler which does couple of things
1. on `optionalFoo.get()` hints that it can be Nullable. 
2. on `optionalFoo.isPresent()` sets the `optionalFoo.get()` access path to Non-null.  

This allows 
```
if (fooOptional.isPresent()) {
  fooOptional.get().bar(); // Implicitly understood to be safe
}
```
and returns a warning if not implied safe. 

Fixes #274 
Added unit tests

### CODE REVIEW

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java [hunk 1]
```diff
@@ -1175,6 +1175,44 @@ public void lambdaPlusRestrictiveNegative() {
...
+        .addSourceLines(
+            "TestPositive.java",
+            "package com.uber;",
+            "import java.util.Optional;",
+            "import javax.annotation.Nullable;",
+            "public class TestPositive {",
+            "  void foo() {",
+            "    Optional<Object> a = Optional.of(null);",
+            "    // BUG: Diagnostic contains: dereferenced expression a.get() is @Nullable",
+            "    a.get().toString();",
```
Comment (Stephan202) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: This doesn't seem right; `Optional.of(null)` will throw a `NullPointerException` because `Optional` can't wrap `null`. So this `a.get()` won't be called, and in other contexts `Optional#get` will never yield `null`.

Or do I misunderstand completely? :thinking:

Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: From what I understand this does not matter. As we are not going to run this code. We just run the static analysis on this code. So as long as we initialise `a` with any method it is alright. 

Anyway if you think I can write this test differently. Let me know.

Comment (Stephan202) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: I agree that in the context of this test no NPE will happen, but the test does show that the plugin would make a "false statement" in case it were applied to real code. After all, `Optional#get()` is never `@Nullable`.

A more realistic test could be:
```java
Optional<Object> a = Optional.empty();"
// BUG: Diagnostic contains: dereferenced expression a.get() may be empty",
a.get().toString();",
```

... but that does require customizing the error message.

Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Thanks 🙂

Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: We should actually be modeling `Optional.of()` such that the parameter is `@NonNull`.  We should add a library model here.

**Comment by msridhar** [comment]
I really like the idea here!  A few high-level thoughts:

* We may need to add modeling for `Optional.of()`, returning an `Optional` whose `get()` result is `@NonNull`.  Not sure how important this is, depends on how many false positives we get.  Also not sure how to implement if we need it.
* We will also get false positives across procedure boundaries (e.g. if you call `isPresent()` in caller and `get()` in the callee).  Again not sure how important it is.  The Checker Framework uses an [`@Present`](https://checkerframework.org/api/org/checkerframework/checker/optional/qual/Present.html) annotation to handle this.
* We may get false positives with Rx streams, if we have a `filter()` checking `isPresent()` and then subsequent callbacks using `get()` without a check.

Bottom line, I like the approach here, but (1) we should almost certainly have it off by default at first, and (2) before we land we should probably test it on some large code base like Uber's to see how many false positives turn up.

@lazaroclapp any further thoughts?

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 1 above)
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Customizing the message might not be trivial. What is the actual result of `Optional<Object> a = Optional.empty(); a.get()` ? An exception?

Comment (Stephan202) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: > What is the actual result of `Optional<Object> a = Optional.empty(); a.get()` ? An exception?

Yep:
```
$ jshell
|  Welcome to JShell -- Version 11.0.2
|  For an introduction type: /help intro

jshell> Optional<Object> a = Optional.empty()
a ==> Optional.empty

jshell> a.get()
|  Exception java.util.NoSuchElementException: No value present
|        at Optional.get (Optional.java:148)
|        at (#2:1)
```

W.r.t. your customization suggestion: SGTM :)

File: nullaway/src/main/java/com/uber/nullaway/Config.java [hunk 2]
```diff
@@ -107,6 +107,9 @@
    */
   boolean acknowledgeRestrictiveAnnotations();
 
+  /** @return true if Optional Emptiness Handler is to be used. */
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/Config.java]: Can we be more specific on what this does?  We should say this enables checking for `Optional.get()` calls without checking `Optional.isPresent()`

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 3]
```diff
@@ -1933,7 +1932,17 @@ private Description matchDereference(
       return Description.NO_MATCH;
     }
     if (mayBeNullExpr(state, baseExpression)) {
-      String message = "dereferenced expression " + baseExpression.toString() + " is @Nullable";
+      String message;
+      if (handler.checkIfOptionalGetCall(baseExpression, state)) {
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/NullAway.java]: I think we want a more general mechanism here, not something specific to `Optional.get()`.  We should refactor and let the handler return a different error message if desired given the `baseExpression` and `state`

File: nullaway/src/main/java/com/uber/nullaway/Config.java (hunk 2 above)
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/Config.java]: Agreed. Either more detail here or maybe even a paragraph or two in the docs and link from here (maybe both?)

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 3 above)
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: Right. We need an entry point into the base `Handler` interface that allows to override error messages for a given combination of message type (see `MessageTypes`), expression, and state.

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java [hunk 4]
```diff
@@ -169,6 +175,23 @@ public NullnessStore getLocalVarInfoBefore(TreePath path, Context context) {
                   || e.getKind().equals(ElementKind.LOCAL_VARIABLE);
             }
           }
+
+          // a filter for Optional get() call
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java]: Why are we adding code to the core here? I thought the idea was to abstract this all behind a handler and guard it with a flag. This code is neither... am I missing something?

File: nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java [hunk 5]
```diff
@@ -270,6 +270,15 @@ NullnessHint onDataflowVisitMethodInvocation(
...
       ExpressionTree tree, NullnessStore thenStore, NullnessStore elseStore);
 
+  /**
+   * Called while creating the error message on a possible null/empty optional deference.
+   *
+   * @param expr The AST node for the expression being matched.
+   * @param state The current visitor state.
+   * @return If the method call in the expression is to Optional.get().
+   */
+  boolean checkIfOptionalGetCall(ExpressionTree expr, VisitorState state);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java]: As above, this should be a more general message overriding method.

File: nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java [hunk 6]
```diff
@@ -289,6 +289,7 @@ private static LibraryModels loadLibraryModels() {
                     "javax.lang.model.util.Elements",
                     "getDocComment(javax.lang.model.element.Element)"),
                 0)
+            .put(methodRef("java.util.Optional", "<T>of(T)"), 0)
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java]: 👍

File: nullaway/src/main/java/com/uber/nullaway/Config.java (hunk 2 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/Config.java]: Done 👍
Will add few lines in the wiki

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 3 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: 👍Done

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java (hunk 4 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java]: Moved it to the handler 👍

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java [hunk 7]
```diff
@@ -176,30 +173,14 @@ public NullnessStore getLocalVarInfoBefore(TreePath path, VisitorState state) {
...
-                      .map(state.getTypes()::erasure);
-              return e.getKind().equals(ElementKind.LOCAL_VARIABLE)
-                  && g.getSimpleName().toString().equals("get")
-                  && g.getParameters().length() == 0
-                  && tbaseType.isPresent()
-                  && state.getTypes().isSubtype(g.owner.type, tbaseType.get());
-            }
-          }
-          return false;
+          return handler.filterApForLocalVarInfoBefore(ap, state);
```
[the remark this answers is not in the document]
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java]: Not sure the name is super clear. But can't also think of a better one. @msridhar ? (This definitely is a better design, though :) )

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java [hunk 8]
```diff
@@ -176,30 +173,14 @@ public NullnessStore getLocalVarInfoBefore(TreePath path, VisitorState state) {
...
         });
   }
 
   /**
    * @param path tree path of static method, or initializer block
    * @param context Javac context
-   * @return fields guaranteed to be nonnull at exit of static method (or initializer block)
+   * @return fields guaranteed to be nonnull at exit of static method (or initialize r block)
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java]: typo?

File: nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java [hunk 9]
```diff
@@ -275,9 +276,21 @@ void onDataflowVisitLambdaResultExpression(
...
-  boolean checkIfOptionalGetCall(ExpressionTree expr, VisitorState state);
+  void getErrorMessage(ExpressionTree expr, VisitorState state, NullAway.ErrorMessage errorMessage);
+
+  /**
+   * Called when the store access paths are filtered for local variable information before an
+   * expression.
+   *
+   * @param accessPath The access path that needs to be checked if filtered.
+   * @param state The current visitor state.
+   * @return true if the accesspath should be filtered to be included in the local variable
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java]: "should be filtered to be included" doesn't easily parse in my mind. Are we excluding or including? (e.g. is this a whitelist or a blacklist?).

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java]: After reading the rest of the code more carefully, I'd go with "@return true if the nullability information for this accesspath should be treated as part of the surrounding context when processing a lambda expression or anonymous class declaration"

I'd also rename  `getLocalVarInfoBefore` to e.g. `getNullnessInfoBeforeNewContext`, to more accurately reflect the fact that it preserves info for APs other than locals. But let's see if that makes sense to @msridhar as well before going ahead with that renaming for a core API method :)

If we do rename that, we can have this method be `Handler.includeApInfoInSavedContext`, or something like that.

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 10]
```diff
@@ -2346,7 +2342,23 @@ public void setComputedNullness(ExpressionTree e, Nullness nullness) {
     UNBOX_NULLABLE,
     NONNULL_FIELD_READ_BEFORE_INIT,
     ANNOTATION_VALUE_INVALID,
-    CAST_TO_NONNULL_ARG_NONNULL;
+    CAST_TO_NONNULL_ARG_NONNULL,
+    GET_ON_EMPTY_OPTIONAL;
+  }
+
+  public static class ErrorMessage {
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: 👍

File: nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java (hunk 9 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java]: I agree with @lazaroclapp's comments and suggested names 😄

**Comment by msridhar** [comment]
LGTM once @lazaroclapp's comments are addressed!

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 11]
```diff
@@ -1932,21 +1933,16 @@ private Description matchDereference(
...
-                + baseExpression.toString().substring(0, exprStringSize - 6)
-                + " can be empty, dereferenced get() call on it";
-      } else {
-        message = "dereferenced expression " + baseExpression.toString() + " is @Nullable";
-      }
+      final String message =
+          "dereferenced expression " + baseExpression.toString() + " is @Nullable";
+      ErrorMessage errorMessage = new ErrorMessage(MessageTypes.DEREFERENCE_NULLABLE, message);
+
+      handler.getErrorMessage(baseExpression, state, errorMessage);
```
[the remark this answers is not in the document]
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: done 👍

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: 👍Done

File: nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java]: 👍Done

File: nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java (hunk 9 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java]: 👍Done

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 12]
```diff
@@ -1960,7 +1961,10 @@ private Description matchDereference(
    * @return the error description
    */
   private Description createErrorDescription(
-      MessageTypes errorType, Tree errorLocTree, String message, TreePath path) {
+      com.uber.nullaway.ErrorMessage.MessageTypes errorType,
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/NullAway.java]: This should be `ErrorMessage.MessageTypes`, no? Rather than the FQN.

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java [hunk 13]
```diff
@@ -1933,11 +1934,16 @@ private Description matchDereference(
...
+      ErrorMessage errorMessage = new ErrorMessage(MessageTypes.DEREFERENCE_NULLABLE, message);
+
+      handler.onPrepareErrorMessage(baseExpression, state, errorMessage);
+
       return createErrorDescriptionForNullAssignment(
-          MessageTypes.DEREFERENCE_NULLABLE,
+          errorMessage.messageType,
           derefExpression,
-          message,
+          errorMessage.message,
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/NullAway.java]: We should probably change all the `createErrorDescription[...]` methods to just take `ErrorMessage` rather than the separate components. But let's leave that for a follow up PR.

Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: Yes, definitely.

File: nullaway/src/main/java/com/uber/nullaway/NullAway.java (hunk 12 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/NullAway.java]: Done 👍
